### PR TITLE
replace clone to clone_shallow and copy only as much as you need attributes.

### PR DESCRIPTION
**Pull request by higumachan**
Resolves: [python-poetry#8250](https://github.com/python-poetry/poetry/issues/8250)

- [X] Added **tests** for changed code.
- [X] Updated **documentation** for changed code.

## Description

This change was made under the assumption that if the tests pass, there should be no problem. However, just replacing the clone with a shallow copy caused several tests to fail, so I made "one level deeper" copies sufficient to fix those tests.

I understand that there is a trade-off with code maintainability regarding this change, so I would appreciate some feedback.

If we are to implement this while ensuring future changeability, I think we could solve this by creating code similar to [Cow (Copy on Write)](https://en.wikipedia.org/wiki/Copy-on-write).

Thanks.

### CODE REVIEW

**Comment by mtkennerly** [comment]
Hi, just cross-posting from https://github.com/python-poetry/poetry/pull/6348#event-10036998403:

I have [a plugin](https://github.com/mtkennerly/poetry-dynamic-versioning) that does VCS-based dynamic versioning, and this would break the handling for dependencies that also use the plugin. If you make this the default, please also add an option for Git dependencies to specify a custom depth, including the full history. For example:

```toml
[tool.poetry.dependencies]
requests = { git = "https://github.com/requests/requests.git", depth = 0 }
```

**Comment by dimbleby** [comment]
@mtkennerly you have misunderstood this MR, it has nothing at all to do with git clone

**Comment by eamanu** [comment]
Did you run a profiler to know if this fix the issue?

File: src/poetry/core/packages/package.py [hunk 1]
```diff
@@ -481,7 +481,8 @@ def without_optional_dependency_groups(self: T) -> T:
         """
         Returns a clone of the package without optional dependency groups.
         """
-        package = self.clone()
+        package = self.clone_shallow()
+        package._dependency_groups = package._dependency_groups.copy()
```
Comment (eamanu) [inline src/poetry/core/packages/package.py:485]: a little comment explaining why we are copying the package._depndency_groups wouldbe great

**Comment by radoering** [comment]
Superseded by #663.

**Pull request closed without merging**